### ISSUE TITLE
Allow configuration of output directory via assemble.

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#39068]
Instead of requiring outputs like `project/macos/Flutter/example/App.framework`, allow some configurable `output/App.framework` path. This could be provided via command line flag or others. This would be distinct from the build directory, which is only a cache for intermediate artifacts.

cc @jmagman

**Comment by jmagman** [comment]
I'd like it to be very explicit and to separate out the logic of assembling out from where the output it.  For example, any command that needs an App framework can instantiate or pass into a AotAssemblyBase subclass a new AotOutputLocations object with properties like:
appUniversalFrameworkDirectory
Map<DarwinArch, Directory> frameworkDirectoryByIOSArchs
etc
That object could inform the output sources?

**Comment by jonahwilliams** [comment]
In your example, does `build framework` (or equivalent) use `frameworkDirectoryByIOSArchs/appUniversalFrameworkDirectory to configure the assemble command?

**Comment by jmagman** [comment]
> In your example, does `build framework` (or equivalent) use `frameworkDirectoryByIOSArchs/appUniversalFrameworkDirectory to configure the assemble command?

I think so.  Right now I'm trying to leverage the logic in  `AotAssemblyBase.build()` to snapshot build and lipo to build the add-to-app iOS frameworks.  Ideally in my new command I could just add AotAssemblyRelease, etc as configured targets since they are close to what I need, or subclass AotAssemblyBase myself with the right inputs and output.  But there's too much opinionation (we really need to make that a word) in `AotAssemblyBase.build()` around what environment variables should be there, where the products get output related to the build directory, etc.

Ditto problems in `DebugBundleFlutterAssets` (though an iOS variant).

So specifying frameworkDirectoryByIOSArchs/appUniversalFrameworkDirectory could be one way to configure it.  Maybe the subclasses need to override it?  Or it gets passed into the target's initialization?  

Similarly, the assemble targets seem opinionated about what variables are populated in the Environment (bitcode, target platform, build mode).  When building frameworks for add-to-app I want to build debug, profile, release build mode frameworks in the same command.  Could these values also be overridden by a subclass/passed in, then the subclass/caller could appropriately check and throw if the expected environment values aren't set?  Or is the expectation for the environment be repopulated between targets?

TL;DR assemble targets should uncouple build logic from output location and command flags to be more modular and re-usable.
